<a href="https://colab.research.google.com/github/estebanramirezcantillo/Esteban-Ramirez---clase-AdN/blob/main/Reto_1_Esteban_Ramirez.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###RETO1 Esteban Ramirez

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
#Cargar datos y filtar
nxl="/content/1. BD2_CarCrash_Int.xlsx"
df=pd.read_excel(nxl)
df = df[df['County'] == 'SAN FRANCISCO']

Introduccion
En este análisis, trabajamos con una base de datos de accidentes automovilísticos en San Francisco, con el objetivo de predecir si un accidente ocurrió en una autopista o no. Para esto, exploramos las variables relevantes y sus características. A continuación, se describe cada una de las variables utilizadas en el modelo:


```
print("\n🔍 Descripción de las Variables:")
print(df_variables.to_string(index=False))
```



In [ ]:
print("\n🔍 Descripción de las Variables:")
print(df_variables.to_string(index=False))


🔍 Descripción de las Variables:
    Variable       Tipo de Dato                                                                       Descripción
     Weekday    Numérica Entera                Día de la semana en que ocurrió el accidente (1=Lunes, 7=Domingo).
    Severity Categórica Binaria                          Nivel de severidad del accidente (0 = Menor, 1 = Mayor).
     ViolCat Categórica Ordinal                                      Clasificación del impacto según su gravedad.
ClearWeather Categórica Binaria Indica si el clima estaba despejado en el momento del accidente (0 = No, 1 = Sí).
       Month    Numérica Entera                   Mes en el que ocurrió el accidente (1 = Enero, 12 = Diciembre).
   CrashType Categórica Nominal           Tipo de accidente (convertido a valor numérico según orden alfabético).
    Daylight  Numérica Continua                               Cantidad de luz diurna en el momento del accidente.
     Highway Categórica Binaria                 Indica 

###1.¿Cuál es el número de datos total de la base de datos?
2081 datos

In [ ]:
variables = ['Weekday', 'Severity', 'ViolCat', 'ClearWeather', 'Month', 'CrashType', 'Daylight', 'Highway']
df = df[variables]

# Convertir 'CrashType' en valores numéricos según orden alfabético
df['CrashType'] = df['CrashType'].astype(str)
df['CrashType'] = df['CrashType'].rank(method='dense').astype(int)


### 2. ¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?

In [ ]:

# Seleccionar solo las variables categóricas
categorical_vars = ['Weekday', 'Severity', 'ViolCat', 'ClearWeather', 'Month', 'CrashType', 'Daylight', 'Highway']
CV = df[categorical_vars]  # Variables de entrada + salida

# Calcular y mostrar los porcentajes de cada categoría dentro de cada variable
for col in CV.columns:
    print(f"Porcentajes para {col}:")
    print(CV[col].value_counts(normalize=True) * 100)
    print("-" * 50)  # Separador visual


Porcentajes para Weekday:
Weekday
3    15.569438
1    14.896684
4    14.608361
5    14.560308
2    14.512254
6    13.407016
7    12.445939
Name: proportion, dtype: float64
--------------------------------------------------
Porcentajes para Severity:
Severity
0    93.70495
1     6.29505
Name: proportion, dtype: float64
--------------------------------------------------
Porcentajes para ViolCat:
ViolCat
3     32.051898
8     10.860163
12    10.091302
10     9.706872
9      9.274387
11     8.505526
7      6.583373
4      5.766458
1      4.853436
6      2.306583
Name: proportion, dtype: float64
--------------------------------------------------
Porcentajes para ClearWeather:
ClearWeather
1    84.718885
0    15.281115
Name: proportion, dtype: float64
--------------------------------------------------
Porcentajes para Month:
Month
4     15.185007
1     11.869294
5     11.821240
3     11.340702
2     11.340702
7     10.475733
6     10.427679
8      6.150889
9      3.459875
10     2.931283
12 

### 3.1 Determinar los intervalos en los cual se encuentra la variación del WEEKDAY para cada una de las categorías de clasificación

In [ ]:

# Definir las variables de clasificación
classification_vars = ['Severity', 'ViolCat', 'ClearWeather', 'Month', 'CrashType', 'Daylight', 'Highway']

# Calcular los intervalos de variación de 'Weekday' para cada variable de clasificación
for var in classification_vars:
    print(f"Intervalos de variación de Weekday para {var}:")
    print(df.groupby(var)['Weekday'].describe()[['min', '25%', '50%', '75%', 'max']])
    print("-" * 50)  # Separador visual


Intervalos de variación de Weekday para Severity:
          min  25%  50%  75%  max
Severity                         
0         1.0  2.0  4.0  6.0  7.0
1         1.0  2.0  4.0  6.0  7.0
--------------------------------------------------
Intervalos de variación de Weekday para ViolCat:
         min   25%  50%   75%  max
ViolCat                           
1        1.0  3.00  6.0  7.00  7.0
3        1.0  2.00  4.0  6.00  7.0
4        1.0  2.00  3.0  5.00  7.0
6        1.0  2.00  3.0  4.25  7.0
7        1.0  2.00  4.0  5.00  7.0
8        1.0  2.00  4.0  5.00  7.0
9        1.0  2.00  4.0  6.00  7.0
10       1.0  2.00  4.0  5.00  7.0
11       1.0  3.00  4.0  6.00  7.0
12       1.0  2.25  4.0  6.00  7.0
--------------------------------------------------
Intervalos de variación de Weekday para ClearWeather:
              min  25%  50%  75%  max
ClearWeather                         
0             1.0  2.0  4.0  6.0  7.0
1             1.0  2.0  4.0  6.0  7.0
-------------------------------------

### 3.2 Determinar los intervalos en los cual se encuentra la variación del ViolCat para cada una de las categorías de clasificación

In [ ]:
# Definir las variables de clasificación
classification_vars2 = ['Severity', 'Weekday', 'ClearWeather', 'Month', 'CrashType', 'Daylight', 'Highway']

# Calcular los intervalos de variación de 'ViolCat' para cada variable de clasificación
for var in classification_vars2:
    print(f"Intervalos de variación de ViolCat para {var}:")
    print(df.groupby(var)['ViolCat'].describe()[['min', '25%', '50%', '75%', 'max']])
    print("-" * 50)  # Separador visual

Intervalos de variación de ViolCat para Severity:
          min  25%  50%   75%   max
Severity                           
0         1.0  3.0  7.0  10.0  12.0
1         1.0  3.0  9.0  11.0  12.0
--------------------------------------------------
Intervalos de variación de ViolCat para Weekday:
         min  25%  50%   75%   max
Weekday                           
1        1.0  3.0  7.0   9.0  12.0
2        1.0  3.0  7.0  10.0  12.0
3        1.0  3.0  8.0  10.0  12.0
4        1.0  3.0  8.0  10.0  12.0
5        1.0  3.0  7.0  10.0  12.0
6        1.0  3.0  7.0  10.0  12.0
7        1.0  3.0  4.0   9.5  12.0
--------------------------------------------------
Intervalos de variación de ViolCat para ClearWeather:
              min  25%  50%   75%   max
ClearWeather                           
0             1.0  3.0  6.0   9.0  12.0
1             1.0  3.0  7.0  10.0  12.0
--------------------------------------------------
Intervalos de variación de ViolCat para Month:
       min  25%  50%    75% 

3.3 Determinar los intervalos en los cual se encuentra la variación del CrashType para cada una de las categorías de clasificación

In [ ]:
# Definir las variables de clasificación
classification_vars2 = ['Severity', 'Weekday', 'ClearWeather', 'Month', 'ViolCat', 'Daylight', 'Highway']

# Calcular los intervalos de variación de 'CrashType' para cada variable de clasificación
for var in classification_vars2:
    print(f"Intervalos de variación de CrashType para {var}:")
    print(df.groupby(var)['CrashType'].describe()[['min', '25%', '50%', '75%', 'max']])
    print("-" * 50)  # Separador visual

Intervalos de variación de CrashType para Severity:
          min  25%  50%  75%  max
Severity                         
0         1.0  3.0  4.0  5.0  7.0
1         1.0  3.0  5.0  7.0  7.0
--------------------------------------------------
Intervalos de variación de CrashType para Weekday:
         min  25%  50%  75%  max
Weekday                         
1        1.0  3.0  3.0  4.0  7.0
2        1.0  3.0  3.0  5.0  7.0
3        1.0  3.0  4.0  5.0  7.0
4        1.0  3.0  4.0  6.0  7.0
5        1.0  3.0  3.0  5.0  7.0
6        1.0  3.0  4.0  5.0  7.0
7        1.0  3.0  4.0  5.0  7.0
--------------------------------------------------
Intervalos de variación de CrashType para ClearWeather:
              min  25%  50%  75%  max
ClearWeather                         
0             1.0  3.0  3.0  5.0  7.0
1             1.0  3.0  4.0  5.0  7.0
--------------------------------------------------
Intervalos de variación de CrashType para Month:
       min  25%  50%   75%  max
Month                 

##Modelo

In [ ]:
# Definir variables independientes (X) y la variable dependiente (Y)
XD = df[['Weekday', 'ViolCat', 'Daylight', 'Month', 'CrashType', 'ClearWeather']]
YD = df[['Highway']]

In [ ]:
from sklearn.naive_bayes import GaussianNB
from re import U
mnb=GaussianNB()
mnb.fit(XD,YD)
std=mnb.var_ #esto es varianza
std=np.sqrt(std)
columns = ['Weekday', 'ViolCat', 'Daylight', 'Month', 'CrashType', 'ClearWeather']
index = ['No Autopista (0)', 'Autopista (1)']
df_std = pd.DataFrame(std, index=index, columns=columns)

# Imprimir tabla con formato
print("\nDesviación estándar por variable y clase de 'Highway':")
print(df_std.to_string(float_format="%.4f"))


Desviación estándar por variable y clase de 'Highway':
                  Weekday  ViolCat  Daylight  Month  CrashType  ClearWeather
No Autopista (0)   1.9506   3.5214    0.4728 2.3434     1.8737        0.3442
Autopista (1)      2.0160   2.1620    0.4852 3.4416     1.0577        0.4012


## Matriz de Confusion

In [ ]:
ydp=mnb.predict(XD)
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(YD,ydp)
# Crear un DataFrame para mostrar la matriz con etiquetas
cm_df = pd.DataFrame(cm, index=["No Autopista (0)", "Autopista (1)"],
                     columns=["Predicho No Autopista (0)", "Predicho Autopista (1)"])

# Imprimir la matriz con formato claro
print("\nMatriz de Confusión:")
print(cm_df)


Matriz de Confusión:
                  Predicho No Autopista (0)  Predicho Autopista (1)
No Autopista (0)                       1365                     215
Autopista (1)                           224                     277


In [ ]:
# Extraer valores de la matriz de confusión
VN = cm[0, 0]  # Verdaderos negativos
FP = cm[0, 1]  # Falsos positivos
FN = cm[1, 0]  # Falsos negativos
VP = cm[1, 1]  # Verdaderos positivos

# Calcular métricas
exactitud = (VN + VP) / (VN + VP + FN + FP)
tasa_error = (FN + FP) / (VN + VP + FN + FP)
sensibilidad = VP / (VP + FN)  # Recall / Sensitivity
especificidad = VN / (VN + FP)
precision_positivos = VP / (VP + FP)
precision_negativos = VN / (VN + FN)
vn_quartiles = np.percentile(VN, [0, 25, 50, 75, 100])
vp_quartiles = np.percentile(VP, [0, 25, 50, 75, 100])

# Resultados Finales

In [ ]:

print("\n🔍 Evaluación del Modelo Naive Bayes:")
print(f"✅ Exactitud: {exactitud:.4f}")
print(f"❌ Tasa de Error: {tasa_error:.4f}")
print(f"📈 Sensibilidad (Recall): {sensibilidad:.4f}")
print(f"📉 Especificidad: {especificidad:.4f}")
print(f"🎯 Precisión Positivos: {precision_positivos:.4f}")
print(f"🎯 Precisión Negativos: {precision_negativos:.4f}")


🔍 Evaluación del Modelo Naive Bayes:
✅ Exactitud: 0.7890
❌ Tasa de Error: 0.2110
📈 Sensibilidad (Recall): 0.5529
📉 Especificidad: 0.8639
🎯 Precisión Positivos: 0.5630
🎯 Precisión Negativos: 0.8590


#Conclusion
El análisis de accidentes automovilísticos en San Francisco utilizando el modelo Naive Bayes permitió evaluar la capacidad de predicción de si un accidente ocurrió en una autopista o no. A partir de las variables seleccionadas, como el día de la semana, nivel de impacto, luz diurna, mes, tipo de choque y clima despejado, se logró construir un modelo de clasificación con un desempeño notable. El modelo alcanzó una exactitud del 78.90%, lo que indica que clasificó correctamente la mayoría de los accidentes. No obstante, la tasa de error fue del 21.10%, reflejando que un porcentaje significativo de las clasificaciones fueron incorrectas. En términos de sensibilidad, el modelo identificó correctamente el 55.29% de los accidentes ocurridos en autopistas, mientras que la especificidad fue del 86.39%, lo que sugiere que el modelo es más preciso en la identificación de accidentes fuera de autopistas que dentro de ellas. La precisión para los positivos fue del 56.30%, lo que significa que de los accidentes clasificados como ocurridos en autopistas, 56.30% realmente lo fueron. Por otro lado, la precisión para los negativos alcanzó 85.90%, indicando que de los accidentes clasificados como no ocurridos en autopistas, 85.90% realmente no lo fueron. En general, los resultados muestran que el modelo Naive Bayes es eficiente para la clasificación de accidentes, con un buen desempeño en la identificación de accidentes fuera de autopistas, pero con margen de mejora en la identificación de accidentes en autopistas. Para optimizar la sensibilidad del modelo, se podrían explorar otras técnicas de preprocesamiento de datos, ajustar los parámetros del modelo o probar algoritmos más avanzados como árboles de decisión o redes neuronales.